### ISSUE TITLE
ValidationError.toString() breaks when forbidNonWhitelisted is enabled in ValidatorOptions

### ISSUE THREAD

**Issue by connorhamilton** [typestack/class-validator#604]
### Description

The [toString()](https://github.com/typestack/class-validator/blob/master/src/validation/ValidationError.ts#L51) method in the validation error class invokes the `map()` method on the `children` property (lines 60 and 70 at commit 1955250). However, this property cannot be guaranteed to be set in all cases. When whitelist and forbidNonWhitelisted are enabled concurrently, the `children` array is set to `undefined` [here](https://github.com/typestack/class-validator/blob/master/src/validation/ValidationExecutor.ts#115). If the `toString()` method is then invoked, a "Cannot ready property 'map' of undefined" error is thrown.

Proposed solution:
Since I don't think the `children` property serves a purpose for a whitelisting error, could the property be updated to be optional in the ValidationError class and a null check added to the `toString()` method before calling `this.children.map()`?

### Reproduction

```typescript
import 'reflect-metadata';

import { validateSync } from 'class-validator';

class Test {
  property: string;
}

const t = new Test();
t.property = "This property has no decorators and should fail validation."

/**
 * Errors are handled as expected since whitelisting is enabled and `property` does not
 * include the @Allow decorator. This produces a single error in the returned array.
 */
const errors = validateSync(t, { whitelist: true, forbidNonWhitelisted: true });
console.log(errors);

/**
 * However, once we try to call the `toString()` method, an error will be thrown since
 * the ValidationError in `errors` doesn't contain a `children` property.
 */
for (const e of errors) {
  console.log(errors.toString());
}
```

### Environment

- [ ] nodejs: 12.16.1
- [ ] browser: 
- [ ] framework/library: 

**class-validator version:** <code>0.12.2</code>

**Comment by giudicelli** [comment]
Hello, 
Can this issue be addressed please, I do have it myself...

Here is a simple code to reproduce.

`import { IsNotEmpty, validateSync } from "class-validator";
import { JsonProperty } from "typescript-json-serializer";

class ClassToCheck {
    @IsNotEmpty({ groups: ["internal"] })
    @JsonProperty()
    public field1: string = "";

    @IsNotEmpty({ groups: ["internal"] })
    @JsonProperty()
    public field2: string = "";
}

const instance = new ClassToCheck();
const errors = validateSync(instance, {
    whitelist: true,
    forbidNonWhitelisted: true,
    groups: ["external"],
});
try {
    if (errors.length) {
        errors.forEach((error) =>
            console.log("Validate error", error.toString())
        );
    }
} catch (toStringError: unknown) {
    console.log("toStringError", toStringError);
}
`

**Comment by simon-a-j** [comment]
I also encountered this and opened PR #846 which should resolve.

There are a couple of other observations I made while addressing this:
1. This is possibly not the only place where we have possibly undefined properties. I think there is scope for some further cleanup on the optional types within the `ValidationError` class, but I think this needs discussion before making changes.
2. I think it might be confusing to refer to the behaviour of `forbidNonWhitelisted` as “throwing” an error since that has a specific meaning in the JS/TS ecosystem. It would be clearer to describe the behaviour as “returning” an error.

Obligatory postscript: I like the project! 👍
